### PR TITLE
ZEN-31948: fix datamap publishing for zing

### PR DESCRIPTION
**Pull request by lundybernard**
fixes ZEN-31948

### CODE REVIEW

File: Products/DataCollector/ApplyDataMap/applydatamap.py [hunk 1]
```diff
@@ -399,6 +403,9 @@ def _process_relationshipmap(relmap, base_device):
         _validate_datamap(parent, object_map)
         for object_map in relmap.maps
     ]
+    for map in new_maps:
+        map.plugin_name = relmap.plugin_name
```
Comment (iancmcc) [inline Products/DataCollector/ApplyDataMap/applydatamap.py]: Although I'm sure this isn't currently happening, probably wise to wrap this in `if not map.plugin_name`

File: Products/DataCollector/plugins/DataMaps.py [hunk 2]
```diff
@@ -58,11 +60,17 @@ def __repr__(self):
...
-        self.maps.append(obj)
+    def append(self, objmap):
+        self._add_map(objmap)
 
     def extend(self, objmaps):
-        self.maps.extend(objmaps)
+        for map in objmaps:
+            self._add_map(map)
+
+    def _add_map(self, objmap):
```
Comment (iancmcc) [inline Products/DataCollector/plugins/DataMaps.py]: Is this necessary in addition to the `plugin_name` that comes in on the rel map?

File: Products/Zing/datamaps.py [hunk 3]
```diff
@@ -192,10 +186,11 @@ def generate_facts(self, zing_tx_state):
         """
         @return: Fact generator
         """
-        log.debug("Processing {} datamaps to send to zing-connector.".format(len(zing_tx_state.datamaps)))
+        print("Processing {} datamaps to send to zing-connector.".format(len(zing_tx_state.datamaps)))
```
Comment (iancmcc) [inline Products/Zing/datamaps.py]: Switch back to `log.debug`

File: Products/Zing/datamaps.py [hunk 4]
```diff
@@ -192,10 +186,11 @@ def generate_facts(self, zing_tx_state):
...
         facts_per_device = defaultdict(list)
         for device, datamap in zing_tx_state.datamaps:
             dm_facts = self.facts_from_datamap(device, datamap, zing_tx_state.datamaps_contexts)
+            print('dm_facts=%s' % dm_facts)
```
Comment (iancmcc) [inline Products/Zing/datamaps.py]: Kill debug code

File: Products/Zing/datamaps.py [hunk 5]
```diff
@@ -243,6 +238,42 @@ def fact_from_object_map(self, om, parent_device=None, relationship=None, contex
...
+            f.metadata["parent"] = idm.parent.getUUID()
+        except Exception:
+            log.debug('parent UUID not found')
+
+        # Hack in whatever extra stuff we need.
+        om_context = (context or {}).get(idm)
+        if om_context is not None:
+            self.apply_extra_fields(om_context, f)
+
+        # FIXME temp solution until we are sure all zenpacks send the plugin
```
Comment (iancmcc) [inline Products/Zing/datamaps.py]: Still necessary?

File: Products/Zing/datamaps.py [hunk 6]
```diff
@@ -255,8 +286,17 @@ def facts_from_datamap(self, device, dm, context):
             f = self.fact_from_object_map(dm, context=context, dm_plugin=dm_plugin)
             if f.is_valid():
                 facts.append(f)
+        elif isinstance(dm, IncrementalDataMap):
+            f = self.fact_from_incremental_map(dm, context=context)
+            if f.is_valid():
+                facts.append(f)
+        else:
+            print('oops datamap type not found')
```
Comment (iancmcc) [inline Products/Zing/datamaps.py]: log.error

File: Products/Zing/datamaps.py [hunk 7]
```diff
@@ -255,8 +286,17 @@ def facts_from_datamap(self, device, dm, context):
...
         return facts
 
+    def facts_from_incrementaldatamap(self):
```
Comment (iancmcc) [inline Products/Zing/datamaps.py]: No. Yes? Empty list?

File: Products/Zing/datamaps.py (hunk 5 above)
Comment (lundybernard) [reply Products/Zing/datamaps.py]: maybe?  this code does not have full test coverage yet, so i'm leaving it as-is

File: Products/Zing/datamaps.py (hunk 6 above)
Comment (lundybernard) [reply Products/Zing/datamaps.py]: done

File: Products/Zing/datamaps.py (hunk 7 above)
Comment (lundybernard) [reply Products/Zing/datamaps.py]: unused method, removed.

File: Products/DataCollector/ApplyDataMap/applydatamap.py (hunk 1 above)
Comment (lundybernard) [reply Products/DataCollector/ApplyDataMap/applydatamap.py]: We considered this, and determined that even if an object map somehow had a plugin_name that differed from the relationship map's, they should use the one from the relmap

File: Products/DataCollector/plugins/DataMaps.py (hunk 2 above)
Comment (lundybernard) [reply Products/DataCollector/plugins/DataMaps.py]: yes, the RelationshipMap and ObjectMap classes should ensure that they contain the correct data.  This covers them so long as datamaps are added through the appropriate API

**Pull request merged**